### PR TITLE
gcc13, gccgo13, gfortran13, gnat13: init at 13.1.0

### PR DESCRIPTION
**Pull request by trofi**
It's a copy of `gcc12.cc` implementation. Nothing special added here.

~Note that gccgo13 does not build yet (similar to existing gccgo12).
The failure is related to libgcc_s.so lookup problems. Not specific
to gcc-13 release.~

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -15133,6 +15177,22 @@ with pkgs;
     meta.broken = stdenv.hostPlatform.isDarwin;
   });
 
+  # Does not build yet:
+  #   ld: warning: libgcc_s.so.1, needed by .../libgo/.libs/libgo.so,
+  #     not found (try using -rpath or -rpath-link)
+  #   libgo.so: undefined reference to `_Unwind_RaiseException@GCC_3.0'
```
Comment (wegank) [inline pkgs/top-level/all-packages.nix:15183]: I think we can still uncomment this, the failure is the same as for `gccgo12`...

Comment (trofi) [reply pkgs/top-level/all-packages.nix]: Sounds reasonable. Uncommented `gccgo13` and updated commit messages.

File: pkgs/development/compilers/gcc/13/default.nix [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+        ../gnat-cflags-11.patch
+        ../gcc-12-gfortran-driving.patch
+        ../ppc-musl.patch
+      ]
+      # We only apply this patch when building a native toolchain for aarch64-darwin, as it breaks building
+      # a foreign one: https://github.com/iains/gcc-12-branch/issues/18
+      ++ optional (stdenv.isDarwin && stdenv.isAarch64 && buildPlatform == hostPlatform && hostPlatform == targetPlatform) (fetchpatch {
+        name = "gcc-12-darwin-aarch64-support.patch";
+        url = "https://github.com/Homebrew/formula-patches/raw/1d184289/gcc/gcc-12.2.0-arm.diff";
+        sha256 = "sha256-omclLslGi/2yCV4pNBMaIpPDMW3tcz/RXdupbNbeOHA=";
```
Comment (wegank) [inline pkgs/development/compilers/gcc/13/default.nix:77]:
```suggestion
        name = "gcc-13-darwin-aarch64-support.patch";
        url = "https://github.com/Homebrew/formula-patches/raw/5c206c47/gcc/gcc-13.1.0.diff";
        sha256 = "sha256-sMgA7nwE2ULa54t5g6VE6eJQYa69XvQrefi9U9f2t4g=";
```

Homebrew/homebrew-core#128954

File: pkgs/development/compilers/gcc/13/default.nix [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+      ++ [
+        ../gnat-cflags-11.patch
+        ../gcc-12-gfortran-driving.patch
+        ../ppc-musl.patch
+      ]
+      # We only apply this patch when building a native toolchain for aarch64-darwin, as it breaks building
+      # a foreign one: https://github.com/iains/gcc-12-branch/issues/18
+      ++ optional (stdenv.isDarwin && stdenv.isAarch64 && buildPlatform == hostPlatform && hostPlatform == targetPlatform) (fetchpatch {
+        name = "gcc-13-darwin-aarch64-support.patch";
+        url = "https://github.com/Homebrew/formula-patches/raw/5c206c47/gcc/gcc-13.1.0.diff";
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/gcc/13/default.nix:76]: Just noticed that we are using a short hash here

```suggestion
        url = "https://github.com/Homebrew/formula-patches/raw/5c206c47e2a08d522ec9795bb314346fff5fc4c5/gcc/gcc-13.1.0.diff";
```

**Review by wegank — APPROVED** [review]
`gcc13` and `gfortran13` work fine on aarch64-darwin.

**Comment by trofi** [comment]
Squashed fixup commits together.

**Comment by wegank** [comment]
@ofborg build gcc13 gcc13.passthru.tests gccgo13 gccgo13.passthru.tests gfortran13 gfortran13.passthru.tests gnat13 gnat13.passthru.tests

**Comment by amjoseph-nixpkgs** [comment]
Built on:
- [x] `powerpc64le-linux`

But:
- Not much luck with cross compiling to just about anything :frowning_face: failures in `libgcc` for `aarch64` and `mips*` hostPlatforms:

```
gcc-aarch64-unknown-linux-gnu> aarch64-unknown-linux-gnu-gcc -O2 -I/nix/store/xgmi380m3z0bq4jjs2yskzmjz50slxjs-glibc-aarch64-unknown-linux-gnu-2.37-8-dev/include -B/nix/store/sv81rcxz0grghyrffjy362c3v14lsd8v-glibc-aarch64-unknown-linux-gnu-2.37-8/lib/ -idirafter /nix/store/xgmi380m3z0bq4jjs2yskzmjz50slxjs-glibc-aarch64-unknown-linux-gnu-2.37-8-dev/include -idirafter /nix/store/dwifmlf7z03dj9axsjm9w7qblyiry939-aarch64-unknown-linux-gnu-stage-final-gcc-12.2.0/lib/gcc/aarch64-unknown-linux-gnu/12.2.0/include-fixed -Wl,-rpath,/nix/store/1sqgwsp628vsrmpz6c7idbrbxp2hn1z0-gcc-aarch64-unknown-linux-gnu-13.1.0-lib/lib -Wl,-L/nix/store/sv81rcxz0grghyrffjy362c3v14lsd8v-glibc-aarch64-unknown-linux-gnu-2.37-8/lib -Wl,-rpath -Wl,/nix/store/sv81rcxz0grghyrffjy362c3v14lsd8v-glibc-aarch64-unknown-linux-gnu-2.37-8/lib -Wl,-dynamic-linker=/nix/store/sv81rcxz0grghyrffjy362c3v14lsd8v-glibc-aarch64-unknown-linux-gnu-2.37-8/lib/ld-linux-aarch64.so.1 --sysroot=/   -O2 -I/nix/store/xgmi380m3z0bq4jjs2yskzmjz50slxjs-glibc-aarch64-unknown-linux-gnu-2.37-8-dev/include -B/nix/store/sv81rcxz0grghyrffjy362c3v14lsd8v-glibc-aarch64-unknown-linux-gnu-2.37-8/lib/ -idirafter /nix/store/xgmi380m3z0bq4jjs2yskzmjz50slxjs-glibc-aarch64-unknown-linux-gnu-2.37-8-dev/include -idirafter /nix/store/dwifmlf7z03dj9axsjm9w7qblyiry939-aarch64-unknown-linux-gnu-stage-final-gcc-12.2.0/lib/gcc/aarch64-unknown-linux-gnu/12.2.0/include-fixed -Wl,-rpath,/nix/store/1sqgwsp628vsrmpz6c7idbrbxp2hn1z0-gcc-aarch64-unknown-linux-gnu-13.1.0-lib/lib -Wl,-L/nix/store/sv81rcxz0grghyrffjy362c3v14lsd8v-glibc-aarch64-unknown-linux-gnu-2.37-8/lib -Wl,-rpath -Wl,/nix/store/sv81rcxz0grghyrffjy362c3v14lsd8v-glibc-aarch64-unknown-linux-gnu-2.37-8/lib -Wl,-dynamic-linker=/nix/store/sv81rcxz0grghyrffjy362c3v14lsd8v-glibc-aarch64-unknown-linux-gnu-2.37-8/lib/ld-linux-aarch64.so.1 -O2  -O2 -I/nix/store/xgmi380m3z0bq4jjs2yskzmjz50slxjs-glibc-aarch64-unknown-linux-gnu-2.37-8-dev/include -B/nix/store/sv81rcxz0grghyrffjy362c3v14lsd8v-glibc-aarch64-unknown-linux-gnu-2.37-8/lib/ -idirafter /nix/store/xgmi380m3z0bq4jjs2yskzmjz50slxjs-glibc-aarch64-unknown-linux-gnu-2.37-8-dev/include -idirafter /nix/store/dwifmlf7z03dj9axsjm9w7qblyiry939-aarch64-unknown-linux-gnu-stage-final-gcc-12.2.0/lib/gcc/aarch64-unknown-linux-gnu/12.2.0/include-fixed -Wl,-rpath,/nix/store/1sqgwsp628vsrmpz6c7idbrbxp2hn1z0-gcc-aarch64-unknown-linux-gnu-13.1.0-lib/lib -Wl,-L/nix/store/sv81rcxz0grghyrffjy362c3v14lsd8v-glibc-aarch64-unknown-linux-gnu-2.37-8/lib -Wl,-rpath -Wl,/nix/store/sv81rcxz0grghyrffjy362c3v14lsd8v-glibc-aarch64-unknown-linux-gnu-2.37-8/lib -Wl,-dynamic-linker=/nix/store/sv81rcxz0grghyrffjy362c3v14lsd8v-glibc-aarch64-unknown-linux-gnu-2.37-8/lib/ld-linux-aarch64.so.1 -DIN_GCC    -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wstrict-prototypes -Wmissing-prototypes -Wold-style-definition  -isystem ./include  -fPIC -g -DIN_LIBGCC2 -fbuilding-libgcc -fno-stack-protector  -fPIC -I. -I. -I../.././gcc -I../../../gcc-13.1.0/libgcc -I../../../gcc-13.1.0/libgcc/. -I../../../gcc-13.1.0/libgcc/../gcc -I../../../gcc-13.1.0/libgcc/../include  -DHAVE_CC_TLS  -Wno-missing-prototypes -Wno-type-limits  -o floattibf.o -MT floattibf.o -MD -MP -MF floattibf.dep  -c ../../../gcc-13.1.0/libgcc/soft-fp/floattibf.c -fvisibility=hidden -DHIDE_EXPORTS
gcc-aarch64-unknown-linux-gnu> In file included from ../../../gcc-13.1.0/libgcc/soft-fp/extendbfsf2.c:31:
gcc-aarch64-unknown-linux-gnu> ../../../gcc-13.1.0/libgcc/soft-fp/brain.h:62:1: error: unable to emulate 'BF'
gcc-aarch64-unknown-linux-gnu>    62 | typedef float BFtype __attribute__ ((mode (BF)));
gcc-aarch64-unknown-linux-gnu>       | ^~~~~~~
gcc-aarch64-unknown-linux-gnu> make[2]: *** [../../../gcc-13.1.0/libgcc/shared-object.mk:14: extendbfsf2.o] Error 1
gcc-aarch64-unknown-linux-gnu> make[2]: *** Waiting for unfinished jobs....
gcc-aarch64-unknown-linux-gnu> In file included from ../../../gcc-13.1.0/libgcc/soft-fp/trunctfbf2.c:30:
gcc-aarch64-unknown-linux-gnu> ../../../gcc-13.1.0/libgcc/soft-fp/brain.h:62:1: error: unable to emulate 'BF'
gcc-aarch64-unknown-linux-gnu>    62 | typedef float BFtype __attribute__ ((mode (BF)));
gcc-aarch64-unknown-linux-gnu>       | ^~~~~~~
gcc-aarch64-unknown-linux-gnu> In file included from ../../../gcc-13.1.0/libgcc/soft-fp/truncsfbf2.c:30:
gcc-aarch64-unknown-linux-gnu> ../../../gcc-13.1.0/libgcc/soft-fp/brain.h:62:1: error: unable to emulate 'BF'
gcc-aarch64-unknown-linux-gnu>    62 | typedef float BFtype __attribute__ ((mode (BF)));
gcc-aarch64-unknown-linux-gnu>       | ^~~~~~~
gcc-aarch64-unknown-linux-gnu> In file included from ../../../gcc-13.1.0/libgcc/soft-fp/trunchfbf2.c:30:
gcc-aarch64-unknown-linux-gnu> ../../../gcc-13.1.0/libgcc/soft-fp/brain.h:62:1: error: unable to emulate 'BF'
gcc-aarch64-unknown-linux-gnu>    62 | typedef float BFtype __attribute__ ((mode (BF)));
gcc-aarch64-unknown-linux-gnu>       | ^~~~~~~
gcc-aarch64-unknown-linux-gnu> In file included from ../../../gcc-13.1.0/libgcc/soft-fp/truncdfbf2.c:30:
gcc-aarch64-unknown-linux-gnu> ../../../gcc-13.1.0/libgcc/soft-fp/brain.h:62:1: error: unable to emulate 'BF'
gcc-aarch64-unknown-linux-gnu>    62 | typedef float BFtype __attribute__ ((mode (BF)));
gcc-aarch64-unknown-linux-gnu>       | ^~~~~~~
gcc-aarch64-unknown-linux-gnu> make[2]: *** [../../../gcc-13.1.0/libgcc/shared-object.mk:14: trunctfbf2.o] Error 1
gcc-aarch64-unknown-linux-gnu> make[2]: *** [../../../gcc-13.1.0/libgcc/shared-object.mk:14: truncsfbf2.o] Error 1
gcc-aarch64-unknown-linux-gnu> make[2]: *** [../../../gcc-13.1.0/libgcc/shared-object.mk:14: truncdfbf2.o] Error 1
gcc-aarch64-unknown-linux-gnu> make[2]: *** [../../../gcc-13.1.0/libgcc/shared-object.mk:14: trunchfbf2.o] Error 1
gcc-aarch64-unknown-linux-gnu> In file included from ../../../gcc-13.1.0/libgcc/soft-fp/floatdibf.c:30:
gcc-aarch64-unknown-linux-gnu> ../../../gcc-13.1.0/libgcc/soft-fp/brain.h:62:1: error: unable to emulate 'BF'
gcc-aarch64-unknown-linux-gnu>    62 | typedef float BFtype __attribute__ ((mode (BF)));
gcc-aarch64-unknown-linux-gnu>       | ^~~~~~~
gcc-aarch64-unknown-linux-gnu> In file included from ../../../gcc-13.1.0/libgcc/soft-fp/floatundibf.c:30:
gcc-aarch64-unknown-linux-gnu> ../../../gcc-13.1.0/libgcc/soft-fp/brain.h:62:1: error: unable to emulate 'BF'
gcc-aarch64-unknown-linux-gnu>    62 | typedef float BFtype __attribute__ ((mode (BF)));
gcc-aarch64-unknown-linux-gnu>       | ^~~~~~~
gcc-aarch64-unknown-linux-gnu> make[2]: *** [../../../gcc-13.1.0/libgcc/shared-object.mk:14: floatdibf.o] Error 1
gcc-aarch64-unknown-linux-gnu> make[2]: *** [../../../gcc-13.1.0/libgcc/shared-object.mk:14: floatundibf.o] Error 1
gcc-aarch64-unknown-linux-gnu> In file included from ../../../gcc-13.1.0/libgcc/soft-fp/floattibf.c:30:
gcc-aarch64-unknown-linux-gnu> ../../../gcc-13.1.0/libgcc/soft-fp/brain.h:62:1: error: unable to emulate 'BF'
gcc-aarch64-unknown-linux-gnu>    62 | typedef float BFtype __attribute__ ((mode (BF)));
gcc-aarch64-unknown-linux-gnu>       | ^~~~~~~
gcc-aarch64-unknown-linux-gnu> make[2]: *** [../../../gcc-13.1.0/libgcc/shared-object.mk:14: floattibf.o] Error 1
gcc-aarch64-unknown-linux-gnu> make[2]: Leaving directory '/build/build/aarch64-unknown-linux-gnu/libgcc'
gcc-aarch64-unknown-linux-gnu> make[1]: *** [Makefile:14026: all-target-libgcc] Error 2
gcc-aarch64-unknown-linux-gnu> make[1]: Leaving directory '/build/build'
gcc-aarch64-unknown-linux-gnu> make: *** [Makefile:1040: all] Error 2
```

**Comment by wegank** [comment]
Same failure for `pkgsCross.aarch64-multiplatform.gcc13` on `x86_64-linux`.

**Comment by wegank** [comment]
@ofborg build pkgsCross.mips64el-linux-gnuabi64.gcc13

**Comment by amjoseph-nixpkgs** [comment]
Yeah I'll deal with mips later, but it would be nice to have cross-to-aarch64 working.

**Comment by trofi** [comment]
Sounds good. I won't be able to get to untangling `pkgsCross` any time soon. Abandoning.

**Comment by amjoseph-nixpkgs** [comment]
The problem is that `gcc12`-based-`stdenv` can't seem to build a `gcc13`-based `crossStageStatic`.

But `gcc13`-based-`stdenv` can!

```
$ nix build -f . -L pkgsCross.aarch64-multiplatform.gcc
```

With:

```patch
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index 03146ca2734c..1138e6f9e5fc 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -14707,7 +14707,7 @@ with pkgs;
   inherit (let
       num =
         if (with stdenv.targetPlatform; isVc4 || libc == "relibc") then 6
-        else 12;
+        else 13;
       numS = toString num;
     in {
       gcc = pkgs.${"gcc${numS}"};
```

Very odd.  Our `bootstrapFiles` (`gcc12`, I think?) can build `xgcc13` but it can't build `crossStageStatic.gcc13`.

**Comment by amjoseph-nixpkgs** [comment]
I guess we really don't have to solve that problem unless some platform has to be held back at `gcc12` for some reason (and is useful as the `buildPlatform` of a cross-compile).

**Comment by vcunat** [comment]
Perhaps we could simply have working gcc13 at least for non-cross for now?

**Comment by trofi** [comment]
If the cross failures are not the concern then it should be good to go.

**Comment by trofi** [comment]
Rebased against current `master` and synced from current `gcc12`, which is:

```diff
--- a/pkgs/development/compilers/gcc/13/default.nix
+++ b/pkgs/development/compilers/gcc/13/default.nix
@@ -29,7 +29,6 @@
 , buildPackages
 , libxcrypt
 , disableGdbPlugin ? !enablePlugin
-, disableBootstrap ? !stdenv.hostPlatform.isDarwin
 , nukeReferences
 , callPackage
 }:
@@ -57,6 +56,7 @@ with builtins;

 let majorVersion = "13";
     version = "${majorVersion}.1.0";
+    disableBootstrap = !stdenv.hostPlatform.isDarwin && !profiledCompiler;

     inherit (stdenv) buildPlatform hostPlatform targetPlatform;

@@ -288,6 +288,8 @@ lib.pipe (stdenv.mkDerivation ({
   targetConfig = if targetPlatform != hostPlatform then targetPlatform.config else null;

   buildFlags =
+    # we do not yet have Nix-driven profiling
+    assert profiledCompiler -> !disableBootstrap;
     let target =
           lib.optionalString (profiledCompiler) "profiled" +
           lib.optionalString (targetPlatform == hostPlatform && hostPlatform == buildPlatform && !disableBootstrap) "bootstrap";
```

**Comment by trofi** [comment]
Had a small chance of looking at

> In file included from ../../../gcc-13.1.0/libgcc/soft-fp/truncdfbf2.c:30:
> ../../../gcc-13.1.0/libgcc/soft-fp/brain.h:62:1: error: unable to emulate 'BF'
>    62 | typedef float BFtype __attribute__ ((mode (BF)));

I think it happens because we use `gcc-12`'s `gcc` to build `gcc-13`'s `libgcc` (which relies on new types present in `gcc-13`). The problem is not specific to `gcc-13`, thus this PR should be fine as is.

To fix cross-builds we should not disable bootstrapping for this case (or should enforce identical compiler version in `$CC`).

**Comment by trofi** [comment]
> I think it happens because we use `gcc-12`'s `gcc` to build `gcc-13`'s `libgcc` (which relies on new types present in `gcc-13`). The problem is not specific to `gcc-13`, thus this PR should be fine as is.
> 
> To fix cross-builds we should not disable bootstrapping for this case (or should enforce identical compiler version in `$CC`).

Somewhat similar problem happens when we try to cross-compile `pkgsCross.riscv64.gcc11.cc`. There `gcc12` with better warnings fails to built 

```
../../../gcc-11.3.0/libgomp/team.c: In function 'gomp_team_start':
../../../gcc-11.3.0/libgomp/team.c:315:34: error: 'start_data' may be used uninitialized [-Werror=maybe-uninitialized]
  315 |   struct gomp_thread_start_data *start_data;
      |                                  ^~~~~~~~~~
```

While `gcc12` should technically be able to build this kind of setup it's never safe to mix major compiler versions like that (due to used builtins, possible ABI changes, possible language defaults changes). Thus I think we should not disable the bootstrap for case of downgrade either.

**Comment by trofi** [comment]
Filed https://github.com/NixOS/nixpkgs/issues/244871 for failing cross-builds of non-default `gcc`.

**Pull request merged**